### PR TITLE
Enable query metrics by default and fixing NPE in GatewayAddressCache

### LINKED ISSUE #14149: [BUG] GatewayAddressCache NPE
```
java.lang.NullPointerException
        at com.azure.cosmos.implementation.directconnectivity.GatewayAddressCache.tryGetAddresses(GatewayAddressCache.java:177)
        at com.azure.cosmos.implementation.directconnectivity.AddressResolver.tryResolveServerPartitionAsync(AddressResolver.java:244)
        at com.azure.cosmos.implementation.directconnectivity.AddressResolver.lambda$resolveAddressesAndIdentityAsync$16(AddressResolver.java:430)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:118)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1782)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:144)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1782)
        at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:241)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1782)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:144)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73)
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1782)
        at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:171)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:838)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:600)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:580)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.onComplete(FluxFlatMap.java:457)
        at reactor.core.publisher.DrainUtils.postCompleteDrain(DrainUtils.java:131)
        at reactor.core.publisher.DrainUtils.postComplete(DrainUtils.java:186)
        at reactor.core.publisher.FluxMapSignal$FluxMapSignalSubscriber.onComplete(FluxMapSignal.java:213)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1783)
        at reactor.core.publisher.MonoCacheTime.subscribeOrReturn(MonoCacheTime.java:140)
        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150)
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1782)
        at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:171)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:838)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:600)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:580)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.onComplete(FluxFlatMap.java:457)
        at reactor.core.publisher.DrainUtils.postCompleteDrain(DrainUtils.java:131)
        at reactor.core.publisher.DrainUtils.postComplete(DrainUtils.java:186)
        at reactor.core.publisher.FluxMapSignal$FluxMapSignalSubscriber.onComplete(FluxMapSignal.java:213)
        at reactor.core.publisher.FluxJust$WeakScalarSubscription.request(FluxJust.java:101)
        at reactor.core.publisher.FluxMapSignal$FluxMapSignalSubscriber.request(FluxMapSignal.java:225)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe(FluxFlatMap.java:363)
        at reactor.core.publisher.FluxMapSignal$FluxMapSignalSubscriber.onSubscribe(FluxMapSignal.java:115)
        at reactor.core.publisher.FluxJust.subscribe(FluxJust.java:70)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4213)
        at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:199)
        at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57)
        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150)
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1782)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:144)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1782)
        at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:171)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:838)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:600)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:580)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.onComplete(FluxFlatMap.java:457)
        at reactor.core.publisher.DrainUtils.postCompleteDrain(DrainUtils.java:131)
        at reactor.core.publisher.DrainUtils.postComplete(DrainUtils.java:186)
        at reactor.core.publisher.FluxMapSignal$FluxMapSignalSubscriber.onComplete(FluxMapSignal.java:213)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1783)
        at reactor.core.publisher.MonoCacheTime.subscribeOrReturn(MonoCacheTime.java:140)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4198)
        at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:199)
        at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57)
        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
        at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:79)
        at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46)
        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150)
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
        at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2344)
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onSubscribe(MonoFlatMap.java:103)
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90)
        at reactor.core.publisher.MonoCurrentContext.subscribe(MonoCurrentContext.java:35)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4213)
        at reactor.core.publisher.Mono.block(Mono.java:1679)
        at com.azure.cosmos.CosmosContainer.blockItemResponse(CosmosContainer.java:223)
        at com.azure.cosmos.CosmosContainer.readItem(CosmosContainer.java:295)
        at site.ycsb.db.AzureCosmosClient.read(AzureCosmosClient.java:281)
        at site.ycsb.DBWrapper.read(DBWrapper.java:135)
        at site.ycsb.workloads.CoreWorkload.doTransactionRead(CoreWorkload.java:722)
        at site.ycsb.workloads.CoreWorkload.doTransaction(CoreWorkload.java:644)
        at site.ycsb.ClientThread.run(ClientThread.java:122)
        at java.lang.Thread.run(Thread.java:748)
        Suppressed: java.lang.Exception: #block terminated with an error
                at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
                at reactor.core.publisher.Mono.block(Mono.java:1680)
```

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
Closes #14149

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java [hunk 1]
```diff
@@ -28,6 +28,7 @@
      * Instantiates a new query request options.
      */
     public CosmosQueryRequestOptions() {
+        this.queryMetricsEnabled = true;
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:31]: how about setting on variable declaration itself

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java [hunk 2]
```diff
@@ -245,9 +255,11 @@ public void queryPlanDiagnostics() {
     }
 
     @Test(groups = {"simple"}, dataProvider = "query", timeOut = TIMEOUT)
-    public void queryMetrics(String query) {
+    public void queryMetrics(String query, Boolean qmEnabled) {
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java:258]: why Boolean vs boolean, here and other place ?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java:258]: To be able to test default (not explicity set)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:31]: Looks like the most common pattern has been to initialize in the ctor if a ctor is present - CosmosDatabaseProperties, CosmosTriggerProperties etc. - so followed that approach for consistency.

**Comment by simplynaveen20** [comment]
@moderakh  Now we are enabling querymetrics by default , thoughts on quick round of perf test ?

**Comment by moderakh** [comment]
> @moderakh Now we are enabling querymetrics by default , thoughts on quick round of perf test ?

Could you please measure the perf impact?
you can use QueryInClauseParallel in 
https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/cosmos/azure-cosmos-benchmark

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java (hunk 1 above)
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:31]: Looks like we didn't update the java doc at first place that the default value is `false`. 

@FabianMeiswinkel  - now would be a good time to update the java docs to mention that default value is `true`

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:31]: Fixed in next iteration.

**Comment by FabianMeiswinkel** [comment]
> Could you please measure the perf impact with/without this PR?
> you can use QueryInClauseParallel in
> https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/cosmos/azure-cosmos-benchmark
> 
> ping me offline I will show you how.

The results of the comparison show minimal regression - as expected - but the variance of the results in pretty high due to the Gateway calls.

QUERY METRICS DISABLED BY DEFAULT
=================================
2020-08-24 23:38:17,710       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [149] seconds.
2020-08-24 23:45:32,670       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [154] seconds.
2020-08-24 23:51:27,380       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [140] seconds.
2020-08-25 00:03:45,282       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [139] seconds.
2020-08-25 00:09:39,985       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [140] seconds.
2020-08-25 00:22:50,043       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [153] seconds.


MIN.: 140 seconds
MAX.: 154 seconds
**AVG.: 145.83 seconds**


QUERY METRICS ENABLED BY DEFAULT
================================
2020-08-24 23:42:10,867       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [150] seconds.
2020-08-24 23:48:29,595       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [145] seconds.
2020-08-24 23:54:22,665       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [147] seconds.
2020-08-25 00:06:35,941       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [144] seconds.
2020-08-25 00:13:30,833       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [137] seconds.
2020-08-25 00:25:47,028       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [20000] operations performed in [153] seconds.


MIN.: 137 seconds
MAX.: 153 seconds
**AVG.: 146 seconds**

**Comment by FabianMeiswinkel** [comment]
Closes #14149

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**